controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/CuratorDb.java

```diff
@@ -376,15 +380,23 @@ public void writeLastRun(Run run) {
     }
 
     public void writeHistoricRuns(ApplicationId id, JobType type, Iterable<Run> runs) {
-        curator.set(runsPath(id, type), asJson(runSerializer.toSlime(runs)));
+        Path path = runsPath(id, type);
+        cachedHistoricRuns.remove(path);
```

Comment:
Normally the `remove` would be placed after `set`, but in this case even the reads are done while holding the same exclusive lock it seems, and in case the ordering doesn't matter.

Reply:
Assume the write fails. Since the cache has been modified with a write that is not reflected in ZK, I thought it better to discard the cache entry here. Writes to the historic runs are only done when a finished run is moved there, and an old entry is (optionally) evicted. This write is retried until it succeeds. 

So, actually, since the runs are indexed by run ID, the cache modification (addition of the new run) would be overwritten again by the next attempt to write the (same) run (with the same run ID) to the cache, so then ordering doesn't matter at all.